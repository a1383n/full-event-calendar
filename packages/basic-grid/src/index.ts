export { DailyGrid } from './lib/dailyGrid'
