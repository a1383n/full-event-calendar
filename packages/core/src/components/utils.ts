import { Event } from '../api/EventImpl'

interface ColList {
  [key: number]: Event[]
}

function createLinesOfColome(eventList: Event[]) {
  const eventListCopy = [...eventList]

  eventListCopy.sort(function (a, b) {
    return new Date(a.start).valueOf() - new Date(b.start).valueOf()
  })

  const colList: ColList = {
    1: []
  }

  function findAvalibaleCol(colNumber: number, event: Event): number {
    if (!colList[colNumber]) {
      colList[colNumber] = []
    }

    const colEvents = colList[colNumber]
    let isColAvalibale = true

    for (let i = 0; i < colEvents.length; i++) {
      if (doEventsOverlap(event, colEvents[i])) {
        isColAvalibale = false
        break
      }
    }

    if (!isColAvalibale) {
      return findAvalibaleCol(colNumber + 1, event)
    }

    return colNumber
  }

  for (let i = 0; i < eventListCopy.length; i++) {
    const num = findAvalibaleCol(1, eventListCopy[i])
    colList[num].push(eventListCopy[i])
  }

  return colList
}

function doEventsOverlap(event1: Event, event2: Event) {
  const start1 = event1.start
  const end1 = event1.end
  const start2 = event2.start
  const end2 = event2.end

  return start1 < end2 && end1 > start2
}

function getEventHeigth(start: Date, end: Date) {
  const minuteCount = end.getHours() * 60 + end.getMinutes() - (start.getHours() * 60 + start.getMinutes())
  const heightInPersentage = (minuteCount / 60) * 100
  return `height:  ${heightInPersentage}%;`
}

function getEventColHeight(top: Event) {
  const eventColHeightInPersentage = (top.start.getHours() + top.start.getMinutes() / 60) * 100
  return `top: ${eventColHeightInPersentage}%;${getEventHeigth(top.start, top.end)}`
}
