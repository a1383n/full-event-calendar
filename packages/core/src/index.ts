export { Calendar } from './Calendar'
export type {CalendarSourceOptions} from './api/CalendarImpl'