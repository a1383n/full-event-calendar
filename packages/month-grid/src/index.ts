export { getExtraRows } from './utils/EventPosition'
export { MonthGrid } from './lib/MonthGrid'
export { addEventsToRows, getMonthRows } from './utils/EventRows'
export { MonthEvent } from './lib/MonthEvent/MonthEvent'
export { useMonthEventDragging } from './utils/EventDragging'
