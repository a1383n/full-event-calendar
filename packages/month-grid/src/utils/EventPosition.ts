import { EventClass } from '@full-event-calendar/shared-ts'
import { ceilDate, floorDate } from '@full-event-calendar/utils'
import { createMemo } from 'solid-js'

export function getLeftPosition(event: EventClass, weekStartDate: Date) {
  const floorWeekStart = floorDate(weekStartDate)
  if (event.start >= floorWeekStart) {
    return event.start.getDay()
  }
  return 0
}

export function getEndPosition(event: EventClass, weekendDate: Date, start: number) {
  if(event.isAllDay && !event.isAllDay()) return 1
  const floorDate = ceilDate(weekendDate)
  if (event.end <= floorDate) {
    return event.end.getDay() - start + 1
  }
  return 6 - start + 1
}

export function leftArrowClass(event: EventClass, weekStartDate: Date, leftArrowClass: boolean) {
  const floorWeekStart = floorDate(weekStartDate)
  if (event.start < floorWeekStart) {
    if (leftArrowClass) {
      return 'clip-path: polygon(100% 0, 10px 0, 0px 50%, 10px 100%, 100% 100%);'
    } else {
      return 'border-top-left-radius:0px;border-bottom-left-radius:0px;left:1px'
    }
  } else {
    return ''
  }
}

export function rightArrowClass(event: EventClass, weekendDate: Date) {
  const floorDate = ceilDate(weekendDate)
  return event.end > floorDate
}

interface eventRows {
  [key: string]: EventClass[]
}

export function getExtraRowsCount(eventRows: eventRows, weekStartDate: Date, weekendDate: Date, rowLimit: number) {
  let rowExtrasCount = [0, 0, 0, 0, 0, 0, 0]
  const arr = Object.keys(eventRows).filter((_, i) => {
    return i + 1 > rowLimit
  })

  for (let index = 0; index < arr.length; index++) {
    const events = eventRows[arr[index]]
    // console.log(events)
    for (let j = 0; j < events.length; j++) {
      const event = events[j]
      const leftP = createMemo(() => getLeftPosition(event, weekStartDate))
      const width = getEndPosition(event, weekendDate, leftP())
      for (let k = leftP(); k < width + leftP(); k++) {
        rowExtrasCount[k] = rowExtrasCount[k] + 1
      }
    }
  }

  return rowExtrasCount
}
