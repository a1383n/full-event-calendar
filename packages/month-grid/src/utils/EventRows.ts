import { EventClass } from '@full-event-calendar/shared-ts'
import { filterEventsByDateRange } from '@full-event-calendar/utils'

export function getMonthRows(res: any, filteredEvents3: any) {
  const finalRes = []
  const finsdis = []

  for (let i = 0; i < res.length; i++) {
    let renges = filterEventsByDateRange(filteredEvents3, res[i][0].date, res[i][6].date)

    finalRes.push(renges)
  }

  for (let i = 0; i < finalRes.length; i++) {
    const mRow = finalRes[i] as EventClass[]
    let rooo = {}
    for (let j = 0; j < mRow.length; j++) {
      const element = mRow[j]
      backRwppaer(rooo, element)
    }
    finsdis.push(rooo)
  }

  return finsdis
}

function backRwppaer(rows: any, event: EventClass) {
  let hasFount = false
  function backTrackRow(rows: any, startP: number, event: EventClass) {
    if (hasFount) return
    if (!rows[startP]) {
      rows[startP] = []
    }
    const arr = rows[startP]
    if (isAvalibleInRow(arr, event)) {
      hasFount = true
      rows[startP].push(event)
    } else {
      backTrackRow(rows, startP + 1, event)
    }
  }
  backTrackRow(rows, 0, event)
}

function isAvalibleInRow(Events: EventClass[], Event: EventClass) {
  let isAvalible = true
  for (let i = 0; i < Events.length; i++) {
    const element = Events[i]
    if (element.checkAllDayOverLap(Event)) {
      isAvalible = false
      break
    }
  }
  return isAvalible
}
