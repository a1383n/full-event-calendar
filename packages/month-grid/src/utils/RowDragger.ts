import { EventClass } from "@full-event-calendar/shared-ts"
import { NewDraggingController, daysDiffInRange } from "@full-event-calendar/utils"


abstract class DraggerHandeler {

  container:HTMLElement;
  constructor(container:HTMLElement){
    this.container = container
  }

}
class RowEditDragger extends DraggerHandeler implements MonthDraggerss {
    draggingController:NewDraggingController|null = null
    private startingDate: Date | null = null
    private currentDate: number | null = null
    onDragStart(event: EventClass,mouseEvent:MouseEvent, startDate?: Date){
      this.draggingController = new NewDraggingController(mouseEvent, event,this.container)
  
      if (startDate) {
        this.startingDate = startDate
      }
  
    }
    onMouseEnter(date: Date,draggerChange:Function) {
      if (this.startingDate) {
        let dayDifference = daysDiffInRange(this.startingDate, date)
        if (this.currentDate != dayDifference) {
          this.currentDate = dayDifference
  
          this.draggingController?.shiftTimeByDay(dayDifference)
          draggerChange()
        }
      } else {
        this.startingDate = date
      }
    }
    onDragEnd(){
      this.startingDate = null
    }
  }


  class RowAddDragger extends DraggerHandeler implements MonthDraggerss {
    draggingController:NewDraggingController|null = null
    private startingDate: Date | null = null
    private currentDate: number | null = null
    onDragStart(event: EventClass,mouseEvent:MouseEvent, startDate?: Date){
      this.draggingController = new NewDraggingController(mouseEvent, event,this.container)
  
      if (startDate) {
        this.startingDate = startDate
      }
  
    }
    onMouseEnter(date: Date,draggerChange:Function) {
      if (this.startingDate) {
        let dayDifference = daysDiffInRange(this.startingDate, date)
        if (this.currentDate != dayDifference) {
          this.currentDate = dayDifference
          if(dayDifference<0){
              this.draggingController?.shiftStartByDay(dayDifference)
              this.draggingController?.shiftEndByDay(0) 
            }else if(dayDifference>0){
              this.draggingController?.shiftEndByDay(dayDifference)
              this.draggingController?.shiftStartByDay(0)

          }else{
            this.draggingController?.shiftEndByDay(dayDifference)
            this.draggingController?.shiftStartByDay(dayDifference)
          }
          draggerChange()
        }
      } else {
        this.startingDate = date
      }
    }
    onDragEnd(){
      this.startingDate = null
    }
  }
  
  interface  MonthDraggerss {
    draggingController:NewDraggingController|null
    onDragStart:(event: EventClass,mouseEvent:MouseEvent, startDate?: Date)=>void
    onMouseEnter:(date: Date,draggerChange:Function)=>void
    onDragEnd:()=>void
  }
  
  export type DraggerTypes = 'editEventRow' | 'addEventRow'
  export class RowDragger {
    dragger : MonthDraggerss
    draggerMode:DraggerTypes
    constructor(draggerMode:DraggerTypes,container:HTMLElement){
        this.draggerMode = draggerMode
        switch (draggerMode) {
            case 'editEventRow':
                this.dragger = new RowEditDragger(container)
                break;
            case 'addEventRow':
                this.dragger = new RowAddDragger(container)
                break;
            default:
                this.dragger = new RowEditDragger(container)
                break;
        }
    }
  }