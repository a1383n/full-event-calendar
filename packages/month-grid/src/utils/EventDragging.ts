import { EventClass } from '@full-event-calendar/shared-ts'
import { NewDraggingController } from '@full-event-calendar/utils'
import { createSignal } from 'solid-js'
import { DraggerTypes, RowDragger } from './RowDragger'

export function useMonthEventDragging(dragEndCb:(event:EventClass,draggerMode:DraggerTypes)=>void,draggerType:DraggerTypes = 'editEventRow') {

  const [draggingEventData, setDraggingEventData] = createSignal<null | NewDraggingController>(null)

  let eventdrager = new RowDragger(draggerType)
  function onDragStart(event: EventClass,mouseEvent:MouseEvent, startDate?: Date) {

    eventdrager.dragger.onDragStart(event,mouseEvent,startDate)
    if (!draggingEventData()) {
      setDraggingEventData(eventdrager.dragger.draggingController)
    }
  }

  function onMouseEnter(date: Date) {
    if (!draggingEventData()) return
    eventdrager.dragger.onMouseEnter(date,()=>{
      setDraggingEventData(null)
      setDraggingEventData(eventdrager.dragger.draggingController)
    })
  }

  function onDragEnd(clearDate:boolean = true) {
    eventdrager.dragger.onDragEnd()
    
    if (!!draggingEventData()) {
      const sourceE = { ...draggingEventData()?.item.sourceEvent }
      // sourceE.start = draggingEventData()?.dragedStartDate as Date
      // sourceE.end = draggingEventData()?.dragedEndDate as Date
      // console.log(draggingEventData())
      sourceE.start = draggingEventData()?.eventSourceStart as Date
      sourceE.end = draggingEventData()?.eventSourceEnd as Date

      if (sourceE) {
        //@ts-ignore
      //  props.onEventUpdate(sourceE)
       dragEndCb(sourceE,eventdrager.draggerMode)
      }
    }

    if(clearDate){
      setDraggingEventData(null)
    }
    
  }

  function changeDraggerType(draggerType:DraggerTypes){
    eventdrager = new RowDragger(draggerType)
  }
  return { onDragEnd, onDragStart, onMouseEnter, draggingEventData ,changeDraggerType ,setDraggingEventData}
}