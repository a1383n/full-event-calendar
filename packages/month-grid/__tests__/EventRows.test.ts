import { expect, test } from 'vitest'
import { getMonthRows } from '../src'
import { events } from '@full-event-calendar/test-events'
import { ArraySplitIntoChunks, sortEventByStart, EventImpl, getCalendarMonthDays } from '@full-event-calendar/utils'

test('getMonthRows', () => {
  const arrList = events.map((ev) => new EventImpl(ev))
  const filteredEvents = sortEventByStart(arrList)
  const monthCalendarObject = getCalendarMonthDays(new Date('Thu Aug 10 2023 15:00:0'), 'persian')

  const monthDateRows = ArraySplitIntoChunks(monthCalendarObject, 7)
  const monthRowGridData = getMonthRows(monthDateRows, filteredEvents)
  expect(monthRowGridData).toMatchSnapshot()
})
