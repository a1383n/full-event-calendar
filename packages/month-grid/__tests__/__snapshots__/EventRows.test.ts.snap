// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`getMonthRows 1`] = `
[
  {},
  {
    "0": [
      EventImpl {
        "duration": 3000,
        "end": 2023-08-03T06:30:00.000Z,
        "id": 16123,
        "name": "some name",
        "sourceEvent": {
          "end": 2023-08-03T06:30:00.000Z,
          "id": 16123,
          "name": "some name",
          "start": 2023-08-01T04:30:00.000Z,
        },
        "start": 2023-08-01T04:30:00.000Z,
      },
      EventImpl {
        "duration": 4440,
        "end": 2023-08-07T06:30:00.000Z,
        "id": 18123,
        "name": "some name",
        "sourceEvent": {
          "end": 2023-08-07T06:30:00.000Z,
          "id": 18123,
          "name": "some name",
          "start": 2023-08-04T04:30:00.000Z,
        },
        "start": 2023-08-04T04:30:00.000Z,
      },
    ],
    "1": [
      EventImpl {
        "duration": 1560,
        "end": 2023-08-02T06:30:00.000Z,
        "id": 17123,
        "name": "some name",
        "sourceEvent": {
          "end": 2023-08-02T06:30:00.000Z,
          "id": 17123,
          "name": "some name",
          "start": 2023-08-01T04:30:00.000Z,
        },
        "start": 2023-08-01T04:30:00.000Z,
      },
      EventImpl {
        "duration": 3000,
        "end": 2023-08-05T06:30:00.000Z,
        "id": 19123,
        "name": "some name",
        "sourceEvent": {
          "end": 2023-08-05T06:30:00.000Z,
          "id": 19123,
          "name": "some name",
          "start": 2023-08-03T04:30:00.000Z,
        },
        "start": 2023-08-03T04:30:00.000Z,
      },
    ],
    "2": [
      EventImpl {
        "duration": 27480,
        "end": 2023-08-20T06:30:00.000Z,
        "id": 30,
        "name": "some name",
        "sourceEvent": {
          "end": 2023-08-20T06:30:00.000Z,
          "id": 30,
          "name": "some name",
          "start": 2023-08-01T04:30:00.000Z,
        },
        "start": 2023-08-01T04:30:00.000Z,
      },
    ],
    "3": [
      EventImpl {
        "duration": 4440,
        "end": 2023-08-04T06:30:00.000Z,
        "id": 33,
        "name": "some name",
        "sourceEvent": {
          "end": 2023-08-04T06:30:00.000Z,
          "id": 33,
          "name": "some name",
          "start": 2023-08-01T04:30:00.000Z,
        },
        "start": 2023-08-01T04:30:00.000Z,
      },
      EventImpl {
        "duration": 36120,
        "end": 2023-08-30T06:30:00.000Z,
        "id": 114,
        "name": "some name",
        "sourceEvent": {
          "end": 2023-08-30T06:30:00.000Z,
          "id": 114,
          "name": "some name",
          "start": 2023-08-05T04:30:00.000Z,
        },
        "start": 2023-08-05T04:30:00.000Z,
      },
    ],
    "4": [
      EventImpl {
        "duration": 4440,
        "end": 2023-08-08T06:30:00.000Z,
        "id": 31,
        "name": "some name",
        "sourceEvent": {
          "end": 2023-08-08T06:30:00.000Z,
          "id": 31,
          "name": "some name",
          "start": 2023-08-05T04:30:00.000Z,
        },
        "start": 2023-08-05T04:30:00.000Z,
      },
    ],
  },
  {
    "0": [
      EventImpl {
        "duration": 27480,
        "end": 2023-08-20T06:30:00.000Z,
        "id": 30,
        "name": "some name",
        "sourceEvent": {
          "end": 2023-08-20T06:30:00.000Z,
          "id": 30,
          "name": "some name",
          "start": 2023-08-01T04:30:00.000Z,
        },
        "start": 2023-08-01T04:30:00.000Z,
      },
    ],
    "1": [
      EventImpl {
        "duration": 4440,
        "end": 2023-08-07T06:30:00.000Z,
        "id": 18123,
        "name": "some name",
        "sourceEvent": {
          "end": 2023-08-07T06:30:00.000Z,
          "id": 18123,
          "name": "some name",
          "start": 2023-08-04T04:30:00.000Z,
        },
        "start": 2023-08-04T04:30:00.000Z,
      },
      EventImpl {
        "duration": 60,
        "end": 2023-08-08T00:30:00.000Z,
        "id": 24,
        "name": "some name",
        "sourceEvent": {
          "end": 2023-08-08T00:30:00.000Z,
          "id": 24,
          "name": "some name",
          "start": 2023-08-07T23:30:00.000Z,
        },
        "start": 2023-08-07T23:30:00.000Z,
      },
      EventImpl {
        "duration": 240,
        "end": 2023-08-09T13:30:00.000Z,
        "id": 21,
        "name": "some name",
        "sourceEvent": {
          "end": 2023-08-09T13:30:00.000Z,
          "id": 21,
          "name": "some name",
          "start": 2023-08-09T09:30:00.000Z,
        },
        "start": 2023-08-09T09:30:00.000Z,
      },
      EventImpl {
        "duration": 1440,
        "end": 2023-08-10T20:29:59.000Z,
        "id": 18,
        "name": "some name",
        "sourceEvent": {
          "end": 2023-08-10T20:29:59.000Z,
          "id": 18,
          "name": "some name",
          "start": 2023-08-09T20:30:00.000Z,
        },
        "start": 2023-08-09T20:30:00.000Z,
      },
      EventImpl {
        "duration": 1440,
        "end": 2023-08-11T20:29:59.000Z,
        "id": 117,
        "name": "some name",
        "sourceEvent": {
          "end": 2023-08-11T20:29:59.000Z,
          "id": 117,
          "name": "some name",
          "start": 2023-08-10T20:30:00.000Z,
        },
        "start": 2023-08-10T20:30:00.000Z,
      },
    ],
    "10": [
      EventImpl {
        "duration": 90,
        "end": 2023-08-08T11:30:00.000Z,
        "id": 25,
        "name": "some name",
        "sourceEvent": {
          "end": 2023-08-08T11:30:00.000Z,
          "id": 25,
          "name": "some name",
          "start": 2023-08-08T10:00:00.000Z,
        },
        "start": 2023-08-08T10:00:00.000Z,
      },
      EventImpl {
        "duration": 120,
        "end": 2023-08-10T14:30:00.000Z,
        "id": 11,
        "name": "some name",
        "sourceEvent": {
          "end": 2023-08-10T14:30:00.000Z,
          "id": 11,
          "name": "some name",
          "start": 2023-08-10T12:30:00.000Z,
        },
        "start": 2023-08-10T12:30:00.000Z,
      },
    ],
    "11": [
      EventImpl {
        "duration": 350,
        "end": 2023-08-10T18:35:00.000Z,
        "id": 12,
        "name": "some name",
        "sourceEvent": {
          "end": 2023-08-10T18:35:00.000Z,
          "id": 12,
          "name": "some name",
          "start": 2023-08-10T12:45:00.000Z,
        },
        "start": 2023-08-10T12:45:00.000Z,
      },
    ],
    "12": [
      EventImpl {
        "duration": 140,
        "end": 2023-08-10T16:00:00.000Z,
        "id": 14,
        "name": "some name",
        "sourceEvent": {
          "end": 2023-08-10T16:00:00.000Z,
          "id": 14,
          "name": "some name",
          "start": 2023-08-10T13:40:00.000Z,
        },
        "start": 2023-08-10T13:40:00.000Z,
      },
    ],
    "13": [
      EventImpl {
        "duration": 165,
        "end": 2023-08-10T16:30:00.000Z,
        "id": 15,
        "name": "some name",
        "sourceEvent": {
          "end": 2023-08-10T16:30:00.000Z,
          "id": 15,
          "name": "some name",
          "start": 2023-08-10T13:45:00.000Z,
        },
        "start": 2023-08-10T13:45:00.000Z,
      },
    ],
    "14": [
      EventImpl {
        "duration": 275,
        "end": 2023-08-10T18:35:00.000Z,
        "id": 13,
        "name": "some name",
        "sourceEvent": {
          "end": 2023-08-10T18:35:00.000Z,
          "id": 13,
          "name": "some name",
          "start": 2023-08-10T14:00:00.000Z,
        },
        "start": 2023-08-10T14:00:00.000Z,
      },
    ],
    "15": [
      EventImpl {
        "duration": 185,
        "end": 2023-08-10T18:35:00.000Z,
        "id": 16,
        "name": "some name",
        "sourceEvent": {
          "end": 2023-08-10T18:35:00.000Z,
          "id": 16,
          "name": "some name",
          "start": 2023-08-10T15:30:00.000Z,
        },
        "start": 2023-08-10T15:30:00.000Z,
      },
    ],
    "16": [
      EventImpl {
        "duration": 185,
        "end": 2023-08-10T19:35:00.000Z,
        "id": 170,
        "name": "some name",
        "sourceEvent": {
          "end": 2023-08-10T19:35:00.000Z,
          "id": 170,
          "name": "some name",
          "start": 2023-08-10T16:30:00.000Z,
        },
        "start": 2023-08-10T16:30:00.000Z,
      },
    ],
    "2": [
      EventImpl {
        "duration": 36120,
        "end": 2023-08-30T06:30:00.000Z,
        "id": 114,
        "name": "some name",
        "sourceEvent": {
          "end": 2023-08-30T06:30:00.000Z,
          "id": 114,
          "name": "some name",
          "start": 2023-08-05T04:30:00.000Z,
        },
        "start": 2023-08-05T04:30:00.000Z,
      },
    ],
    "3": [
      EventImpl {
        "duration": 4440,
        "end": 2023-08-08T06:30:00.000Z,
        "id": 31,
        "name": "some name",
        "sourceEvent": {
          "end": 2023-08-08T06:30:00.000Z,
          "id": 31,
          "name": "some name",
          "start": 2023-08-05T04:30:00.000Z,
        },
        "start": 2023-08-05T04:30:00.000Z,
      },
      EventImpl {
        "duration": 90,
        "end": 2023-08-09T11:30:00.000Z,
        "id": 22,
        "name": "some name",
        "sourceEvent": {
          "end": 2023-08-09T11:30:00.000Z,
          "id": 22,
          "name": "some name",
          "start": 2023-08-09T10:00:00.000Z,
        },
        "start": 2023-08-09T10:00:00.000Z,
      },
      EventImpl {
        "duration": 2400,
        "end": 2023-08-11T12:29:59.000Z,
        "id": 19,
        "name": "some name",
        "sourceEvent": {
          "end": 2023-08-11T12:29:59.000Z,
          "id": 19,
          "name": "some name",
          "start": 2023-08-09T20:30:00.000Z,
        },
        "start": 2023-08-09T20:30:00.000Z,
      },
    ],
    "4": [
      EventImpl {
        "duration": 4440,
        "end": 2023-08-09T06:30:00.000Z,
        "id": 110,
        "name": "some name",
        "sourceEvent": {
          "end": 2023-08-09T06:30:00.000Z,
          "id": 110,
          "name": "some name",
          "start": 2023-08-06T04:30:00.000Z,
        },
        "start": 2023-08-06T04:30:00.000Z,
      },
      EventImpl {
        "duration": 125,
        "end": 2023-08-10T10:35:00.000Z,
        "id": 6,
        "name": "some name",
        "sourceEvent": {
          "end": 2023-08-10T10:35:00.000Z,
          "id": 6,
          "name": "some name",
          "start": 2023-08-10T08:30:00.000Z,
        },
        "start": 2023-08-10T08:30:00.000Z,
      },
    ],
    "5": [
      EventImpl {
        "duration": 1560,
        "end": 2023-08-07T06:30:00.000Z,
        "id": 111,
        "name": "some name",
        "sourceEvent": {
          "end": 2023-08-07T06:30:00.000Z,
          "id": 111,
          "name": "some name",
          "start": 2023-08-06T04:30:00.000Z,
        },
        "start": 2023-08-06T04:30:00.000Z,
      },
      EventImpl {
        "duration": 180,
        "end": 2023-08-08T04:29:59.000Z,
        "id": 186,
        "name": "some name",
        "sourceEvent": {
          "end": 2023-08-08T04:29:59.000Z,
          "id": 186,
          "name": "some name",
          "start": 2023-08-08T01:30:00.000Z,
        },
        "start": 2023-08-08T01:30:00.000Z,
      },
      EventImpl {
        "duration": 120,
        "end": 2023-08-09T13:30:00.000Z,
        "id": 20,
        "name": "some name",
        "sourceEvent": {
          "end": 2023-08-09T13:30:00.000Z,
          "id": 20,
          "name": "some name",
          "start": 2023-08-09T11:30:00.000Z,
        },
        "start": 2023-08-09T11:30:00.000Z,
      },
      EventImpl {
        "duration": 240,
        "end": 2023-08-10T13:30:00.000Z,
        "id": 8,
        "name": "some name",
        "sourceEvent": {
          "end": 2023-08-10T13:30:00.000Z,
          "id": 8,
          "name": "some name",
          "start": 2023-08-10T09:30:00.000Z,
        },
        "start": 2023-08-10T09:30:00.000Z,
      },
    ],
    "6": [
      EventImpl {
        "duration": 1560,
        "end": 2023-08-09T06:30:00.000Z,
        "id": 112,
        "name": "some name",
        "sourceEvent": {
          "end": 2023-08-09T06:30:00.000Z,
          "id": 112,
          "name": "some name",
          "start": 2023-08-08T04:30:00.000Z,
        },
        "start": 2023-08-08T04:30:00.000Z,
      },
      EventImpl {
        "duration": 90,
        "end": 2023-08-10T11:30:00.000Z,
        "id": 9,
        "name": "some name",
        "sourceEvent": {
          "end": 2023-08-10T11:30:00.000Z,
          "id": 9,
          "name": "some name",
          "start": 2023-08-10T10:00:00.000Z,
        },
        "start": 2023-08-10T10:00:00.000Z,
      },
    ],
    "7": [
      EventImpl {
        "duration": 1560,
        "end": 2023-08-09T06:30:00.000Z,
        "id": 113,
        "name": "some name",
        "sourceEvent": {
          "end": 2023-08-09T06:30:00.000Z,
          "id": 113,
          "name": "some name",
          "start": 2023-08-08T04:30:00.000Z,
        },
        "start": 2023-08-08T04:30:00.000Z,
      },
      EventImpl {
        "duration": 150,
        "end": 2023-08-10T13:00:00.000Z,
        "id": 10,
        "name": "some name",
        "sourceEvent": {
          "end": 2023-08-10T13:00:00.000Z,
          "id": 10,
          "name": "some name",
          "start": 2023-08-10T10:30:00.000Z,
        },
        "start": 2023-08-10T10:30:00.000Z,
      },
    ],
    "8": [
      EventImpl {
        "duration": 7320,
        "end": 2023-08-13T06:30:00.000Z,
        "id": 32,
        "name": "some name",
        "sourceEvent": {
          "end": 2023-08-13T06:30:00.000Z,
          "id": 32,
          "name": "some name",
          "start": 2023-08-08T04:30:00.000Z,
        },
        "start": 2023-08-08T04:30:00.000Z,
      },
    ],
    "9": [
      EventImpl {
        "duration": 120,
        "end": 2023-08-08T06:30:00.000Z,
        "id": 23,
        "name": "some name",
        "sourceEvent": {
          "end": 2023-08-08T06:30:00.000Z,
          "id": 23,
          "name": "some name",
          "start": 2023-08-08T04:30:00.000Z,
        },
        "start": 2023-08-08T04:30:00.000Z,
      },
      EventImpl {
        "duration": 120,
        "end": 2023-08-10T13:30:00.000Z,
        "id": 7,
        "name": "some name",
        "sourceEvent": {
          "end": 2023-08-10T13:30:00.000Z,
          "id": 7,
          "name": "some name",
          "start": 2023-08-10T11:30:00.000Z,
        },
        "start": 2023-08-10T11:30:00.000Z,
      },
    ],
  },
  {
    "0": [
      EventImpl {
        "duration": 27480,
        "end": 2023-08-20T06:30:00.000Z,
        "id": 30,
        "name": "some name",
        "sourceEvent": {
          "end": 2023-08-20T06:30:00.000Z,
          "id": 30,
          "name": "some name",
          "start": 2023-08-01T04:30:00.000Z,
        },
        "start": 2023-08-01T04:30:00.000Z,
      },
    ],
    "1": [
      EventImpl {
        "duration": 36120,
        "end": 2023-08-30T06:30:00.000Z,
        "id": 114,
        "name": "some name",
        "sourceEvent": {
          "end": 2023-08-30T06:30:00.000Z,
          "id": 114,
          "name": "some name",
          "start": 2023-08-05T04:30:00.000Z,
        },
        "start": 2023-08-05T04:30:00.000Z,
      },
    ],
    "2": [
      EventImpl {
        "duration": 7320,
        "end": 2023-08-13T06:30:00.000Z,
        "id": 32,
        "name": "some name",
        "sourceEvent": {
          "end": 2023-08-13T06:30:00.000Z,
          "id": 32,
          "name": "some name",
          "start": 2023-08-08T04:30:00.000Z,
        },
        "start": 2023-08-08T04:30:00.000Z,
      },
      EventImpl {
        "duration": 1620,
        "end": 2023-08-20T04:29:59.000Z,
        "id": 189,
        "name": "some name",
        "sourceEvent": {
          "end": 2023-08-20T04:29:59.000Z,
          "id": 189,
          "name": "some name",
          "start": 2023-08-19T01:30:00.000Z,
        },
        "start": 2023-08-19T01:30:00.000Z,
      },
    ],
  },
  {
    "0": [
      EventImpl {
        "duration": 27480,
        "end": 2023-08-20T06:30:00.000Z,
        "id": 30,
        "name": "some name",
        "sourceEvent": {
          "end": 2023-08-20T06:30:00.000Z,
          "id": 30,
          "name": "some name",
          "start": 2023-08-01T04:30:00.000Z,
        },
        "start": 2023-08-01T04:30:00.000Z,
      },
    ],
    "1": [
      EventImpl {
        "duration": 36120,
        "end": 2023-08-30T06:30:00.000Z,
        "id": 114,
        "name": "some name",
        "sourceEvent": {
          "end": 2023-08-30T06:30:00.000Z,
          "id": 114,
          "name": "some name",
          "start": 2023-08-05T04:30:00.000Z,
        },
        "start": 2023-08-05T04:30:00.000Z,
      },
    ],
    "2": [
      EventImpl {
        "duration": 1620,
        "end": 2023-08-20T04:29:59.000Z,
        "id": 189,
        "name": "some name",
        "sourceEvent": {
          "end": 2023-08-20T04:29:59.000Z,
          "id": 189,
          "name": "some name",
          "start": 2023-08-19T01:30:00.000Z,
        },
        "start": 2023-08-19T01:30:00.000Z,
      },
    ],
  },
]
`;
